tdrs-frontend/src/components/IdleTimer/IdleTimer.jsx

```diff
@@ -0,0 +1,123 @@
+import React, { useRef, useState, useEffect } from 'react'
+import { useIdleTimer } from 'react-idle-timer'
+import { useDispatch } from 'react-redux'
+import Button from '../Button'
+import { fetchAuth } from '../../actions/auth'
+
+/**
+ * IdleTimer uses the `react-idle-timer` package to watch for user inactivity.
+ * There are environment variables set for the timeout time (how long before the
+ * inactivity modal pops up) and the time between when the modal pops up and the
+ * user is automatically logged out.
+ * There is a second variable for the debounce time to minimize calls to
+ * `/v1/auth_check` on user activity.
+ */
+function IdleTimer() {
+  const [isModalVisible, setIsModalVisible] = useState(false)
+  const dispatch = useDispatch()
+
+  useEffect(() => {
+    function keyListener(e) {
+      const listener = keyListenersMap.get(e.keyCode)
+      return listener && listener(e)
+    }
+    document.addEventListener('keydown', keyListener)
+
+    return () => document.removeEventListener('keydown', keyListener)
+  })
+
+  useEffect(() => {
+    let timer
+    if (isModalVisible) {
+      headerRef.current.focus()
+      timer = setTimeout(onSignOut, process.env.REACT_APP_LOGOUT_TIME)
+    }
+
+    return () => clearTimeout(timer)
+  })
+
+  const onSignOut = () => {
+    window.location.href = `${process.env.REACT_APP_BACKEND_URL}/logout/oidc`
+  }
+
+  const onRenewSession = () => {
+    dispatch(fetchAuth())
+    setIsModalVisible(false)
+  }
+
+  const modalRef = useRef()
+  const headerRef = useRef()
+  const handleTabKey = (e) => {
+    if (isModalVisible) {
+      const focusableModalElements = modalRef.current.querySelectorAll('button')
+      const firstElement = focusableModalElements[0]
+      const lastElement =
+        focusableModalElements[focusableModalElements.length - 1]
+
+      if (!e.shiftKey && document.activeElement !== firstElement) {
+        firstElement.focus()
+        return e.preventDefault()
+      }
+
+      if (e.shiftKey && document.activeElement !== lastElement) {
+        lastElement.focus()
+        e.preventDefault()
+      }
+    }
+
+    return null
+  }
+
+  const keyListenersMap = new Map([
+    [27, onRenewSession],
+    [9, handleTabKey],
+  ])
+
+  useIdleTimer({
+    timeout: process.env.REACT_APP_TIMEOUT_TIME,
+    onIdle: () => {
+      setIsModalVisible(true)
+    },
+    onAction: () => {
+      if (!isModalVisible) {
+        dispatch(fetchAuth())
+      }
+    },
+    debounce: process.env.REACT_APP_DEBOUNCE_TIME,
+  })
```

Comment:
This is the third party package that I'm using for the session timeout. The timeout variables come from the environment variables I defined. `Timeout` is how long before the modal pops up. `onAction` is called when there is an action and (since any action is sending a request to renew the auth token) those requests are debounced by the `debounce` key.